ops/c/src/externlib/ops_hdf5_common.cpp

```diff
@@ -82,4 +82,165 @@ void split_h5_name(const char *data_name,
   while (std::getline(name_stream, segment, '/')) {
     h5_name_list.push_back(segment);
   }
+}
+
+hid_t create_float16_type() {
+  hid_t new_type = H5Tcopy(H5T_IEEE_F32LE);
+  size_t spos, epos, esize, mpos, msize;
+  H5Tget_fields(new_type, &spos, &epos, &esize, &mpos, &msize);
+  // for float16
+  mpos = 0;
+  msize = 10;
+  esize = 5;
+  epos = 10;
+  spos = 15;
+  // for single precision
+  //   mpos = 0;
+  // msize = 23;
+  // esize = 8;
+  // epos = 23;
+  // spos = 31;
+  H5Tset_fields(new_type, spos, epos, esize, mpos, msize);
+  H5Tset_precision(new_type, 16);
+  H5Tset_size(new_type, 2);
+  H5Tset_ebias(new_type, 15);
+  return new_type;
+}
+  // create the dataset or open the dataset if existing
+void H5_dataset_space(const hid_t file_id, const int data_dims,
+                      const hsize_t *global_data_size,
+                      const std::vector<std::string> &h5_name_list,
+                      const char *data_type, const int real_precision,
+                      std::vector<hid_t> &groupid_list, hid_t &dataset_id,
+                      hid_t &file_space) {
+
+  hid_t parent_group{file_id};
+  const char *data_name = h5_name_list.back().c_str();
+  for (int grp = 0; grp < (h5_name_list.size() - 1); grp++) {
+    if (H5Lexists(parent_group, h5_name_list[grp].c_str(), H5P_DEFAULT) == 0) {
+      parent_group = H5Gcreate(parent_group, h5_name_list[grp].c_str(),
+                               H5P_DEFAULT, H5P_DEFAULT, H5P_DEFAULT);
+    } else {
+      parent_group =
+          H5Gopen(parent_group, h5_name_list[grp].c_str(), H5P_DEFAULT);
+    }
+    groupid_list[grp] = parent_group;
+  }
+
+  if (H5Lexists(parent_group, data_name, H5P_DEFAULT) == 0) {
+    hid_t data_plist_id{H5Pcreate(H5P_DATASET_CREATE)};
+    file_space = H5Screate_simple(data_dims, global_data_size, NULL);
+
+    hid_t type = h5_type(data_type);
+    if (type == H5T_NATIVE_FLOAT || type == H5T_NATIVE_DOUBLE) {
+      hid_t real_type;
+      bool float16_created{false};
+      if (type == H5T_NATIVE_DOUBLE) {
+        if (real_precision == 16) {
```

Comment:
what is 4, 8, 16 here? 4 is half precision? that is very misleading, because sizeof(float) = 4

Reply:
They are intended for useful figures of a real number at various precision. A kind of Fortran way adopted (i.e., selected_real_kind) here. Happy to move to other conventions, let's do 2 for half, 4 for float and 8 for double?

Reply:
The HDF5 convention is 16(half) 32(single) 64(double), we can probably use as well

Reply:
We might define some  enumeration for this.

Reply:
either bits - then it's 16,32,64, or bytes - 2,4,8. Yes, and enum would be good for this!